### ISSUE TITLE
selector in nested matrix generators not regarded

### ISSUE THREAD

**Issue by Roshick** [argoproj/argo-cd#11982]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

If using a nested matrix generator, the selectors of the nested generators are not regarded in any way.

**To Reproduce**

Setup:

https://github.com/Roshick/poc-payload
https://github.com/Roshick/poc-workload

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: test-set
spec:
  generators:
  - matrix:
      generators:
      - git:
          repoURL: https://github.com/Roshick/poc-platform.git
          revision: HEAD
          pathParamPrefix: clusterCfg
          files:
          - path: clusters/*.yaml
      - matrix:
          generators:
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: serviceCfg
              files:
              - path: services/*/service.yaml
            selector:
              matchExpressions:
              - {key: providers.test-set-foo, operator: Exists}
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: envCfg
              files:
              - path: services/{{service}}/dev.yaml
  template:
    metadata:
      name: '{{service}}-dev-{{cluster}}'
      labels:
        environment.type: dev
        service: '{{service}}'
    spec:
      project: default
      source:
        repoURL: '{{repoURL}}'
        targetRevision: '{{targetRevision}}'
        chart: '{{chart}}'
      destination:
        server: https://kubernetes.default.svc
        namespace: default
```

**Expected behavior**

As https://github.com/Roshick/poc-workload/blob/main/services/service-a/service.yaml does not contain providers.test-set-foo, the inner git generator in the nested matrix generator should not produce any values and the application should not be created.

Setting the selector to the outer matrix generator produces the expected results:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: test-set
spec:
  generators:
  - matrix:
      generators:
      - git:
          repoURL: https://github.com/Roshick/poc-platform.git
          revision: HEAD
          pathParamPrefix: clusterCfg
          files:
          - path: clusters/*.yaml
      - matrix:
          generators:
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: serviceCfg
              files:
              - path: services/*/service.yaml
          - git:
              repoURL: https://github.com/Roshick/poc-workload.git
              revision: HEAD
              pathParamPrefix: envCfg
              files:
              - path: services/{{service}}/dev.yaml
        selector:
          matchExpressions:
          - {key: providers.test-set-foo, operator: Exists}
  template:
    metadata:
      name: '{{service}}-dev-{{cluster}}'
      labels:
        environment.type: dev
        service: '{{service}}'
    spec:
      project: default
      source:
        repoURL: '{{repoURL}}'
        targetRevision: '{{targetRevision}}'
        chart: '{{chart}}'
      destination:
        server: https://kubernetes.default.svc
        namespace: default
```

**Version**

2.6.0-rc2

**Comment by Roshick** [comment]
I think I have located the issue, just need my local setup running to test my fix.

**Comment by crenshaw-dev** [comment]
The fix will be in 2.8.0-rc1. I won't be cherry-picking back since the fix was pretty involved.
